ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/application/ApplicationPluginBuilder.kt

```diff
@@ -82,151 +49,171 @@ public abstract class PluginBuilder<PluginConfig : Any> internal constructor(
 
     internal val pipelineHandlers: MutableList<PipelineHandler> = mutableListOf()
 
-    internal fun newPhase(): PipelinePhase = PipelinePhase("${key.name}Phase${Random.nextInt()}")
-
-    private fun <T : Any, ContextT : CallHandlingContext> onDefaultPhaseWithMessage(
-        interceptions: MutableList<Interception<T>>,
-        phase: PipelinePhase,
-        handlerName: String,
-        contextInit: (PipelineContext<T, ApplicationCall>) -> ContextT,
-        block: suspend ContextT.(ApplicationCall, Any) -> Unit
-    ) {
-        interceptions.add(
-            Interception(
-                phase,
-                action = { pipeline ->
-                    pipeline.intercept(phase) {
-                        // Information about the plugin name is needed for Intellij Idea debugger.
-                        addToContextInDebugMode(PluginName(key.name)) {
-                            ijDebugReportHandlerStarted(pluginName = key.name, handler = handlerName)
-
-                            // Perform current plugin's handler
-                            contextInit(this@intercept).block(call, subject)
-
-                            ijDebugReportHandlerFinished(pluginName = key.name, handler = handlerName)
-                        }
-                    }
-                }
-            )
-        )
-    }
-
-    private fun <T : Any, ContextT : CallHandlingContext> onDefaultPhase(
-        interceptions: MutableList<Interception<T>>,
-        phase: PipelinePhase,
-        handlerName: String,
-        contextInit: (PipelineContext<T, ApplicationCall>) -> ContextT,
-        block: suspend ContextT.(ApplicationCall) -> Unit
-    ) {
-        onDefaultPhaseWithMessage(interceptions, phase, handlerName, contextInit) { call, _ -> block(call) }
-    }
+    internal val events: MutableList<HookHandler<*>> = mutableListOf()
 
     /**
-     * Specifies how to modify HTTP call handling for the current [PluginBuilder].
-     * @see OnCall
+     * Defines how processing an HTTP call needs to be modified by the current [ApplicationPluginBuilder].
+     *
+     * @param block An action that needs to be executed when your application receives an HTTP call.
      **/
-    public override val onCall: OnCall = object : OnCall {
-        private val plugin get() = this@PluginBuilder
-
-        override operator fun invoke(block: suspend CallContext.(ApplicationCall) -> Unit) {
-            plugin.onDefaultPhase(
-                plugin.callInterceptions,
-                ApplicationCallPipeline.Plugins,
-                PHASE_ON_CALL,
-                ::CallContext
-            ) { call ->
-                block(call)
-            }
+    public override fun onCall(block: suspend OnCallContext<PluginConfig>.(call: ApplicationCall) -> Unit) {
```

Comment:
it's a little inconsistent, that we have `OnCallContext` but `CallReceiveContext` and `CallRespondContext`

Reply:
Yep, it looks like it should be `OnCallReceiveContext`